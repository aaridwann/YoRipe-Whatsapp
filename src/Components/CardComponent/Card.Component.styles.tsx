import { StyleSheet } from "react-native";

import { scaleHeight, scaleSize, scaleWidth } from "../../Utils/Size/size.utils";

const styles = StyleSheet.create({
  container: {
    marginVertical: scaleHeight(4),
    width: scaleWidth(320),
    minHeight: scaleHeight(65),
    backgroundColor: "white",
    borderRadius: 8,
    padding: scaleSize(8),
    overflow: "hidden",
    flexDirection: "row",
    justifyContent: "space-between",
  },
  iconMarked: {
    position: "absolute",
    top: scaleHeight(-5),
    zIndex: 1,
    right: 0,
  },
  iconEdit: { marginTop: scaleHeight(10), margin: scaleSize(2) },
  toolsWrapper: { flexDirection: "row" },
  iconDelete: { marginTop: scaleHeight(10), margin: scaleSize(2) },
  titleText: { fontWeight: "700" },
  textWrapper: { height: "100%" },
  descriptionText: { marginTop: scaleHeight(5), textAlign: "left" },
});

export default styles;
