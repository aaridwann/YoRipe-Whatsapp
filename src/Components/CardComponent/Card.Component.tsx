import React from 'react';
import { Text, TouchableOpacity, View } from 'react-native';
import Ionicons from '@expo/vector-icons/Ionicons';

import { scaleHeight, scaleSize } from '../../Utils/Size/size.utils';
import styles from './Card.Component.styles';
import { Props } from './Card.Component.types';

const ellipsisText = (string: string, limit?: number): string => {
    const limitLength = limit || 100;
    return string.length > limitLength
        ? string.substring(0, limitLength - 3) + '...'
        : string;
};

const _renderDescription = (text: string): React.ReactNode => {
    return <Text style={styles.descriptionText}>{ellipsisText(text, 100)}</Text>;
};

const _renderText = (props: Props) => (
    <View style={styles.textWrapper}>
        <Text
            accessibilityLabel={props.accessibilityLabel}
            style={styles.titleText}
        >
            {props.title}
        </Text>
        {_renderDescription(props.description)}
    </View>
);

const _renderDeleteIcon = (props: Props) => (
    <TouchableOpacity
        accessibilityLabel={`${props.accessibilityLabel}_Delete_Button`}
        onPress={props.onDelete}
    >
        <Ionicons
            style={styles.iconDelete}
            name="trash-outline"
            size={scaleSize(20)}
            color="gray"
        />
    </TouchableOpacity>
);

const _renderEditIcon = (props: Props) => (
    <TouchableOpacity
        accessibilityLabel={`${props.accessibilityLabel}_Edit_Button`}
        onPress={props.onEdit}
    >
        <Ionicons
            style={styles.iconEdit}
            name="pencil-outline"
            size={scaleSize(20)}
            color="gray"
        />
    </TouchableOpacity>
);

const _renderTools = (props: Props): React.ReactNode => (
    <View style={styles.toolsWrapper}>
        {_renderEditIcon(props)}
        {_renderDeleteIcon(props)}
    </View>
);

const _renderMarked = (label: string): React.ReactNode => (
    <Ionicons
        accessibilityLabel={`${label}_Marked_Icon`}
        style={styles.iconMarked}
        name="md-checkmark-circle"
        size={scaleSize(20)}
        color="green"
    />
);

const _renderContent = (props: Props) => (
    <View style={styles.container}>
        {_renderText(props)}
        {_renderTools(props)}
    </View>
);

const CardComponent: React.FC<Props> = (props) => (
    <View style={{ marginVertical: scaleHeight(5) }}>
        {props.isMarked && _renderMarked(props.accessibilityLabel)}
        {_renderContent(props)}
    </View>
);

export default CardComponent;
