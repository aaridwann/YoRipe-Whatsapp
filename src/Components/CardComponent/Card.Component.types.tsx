export interface Props {
  title: string;
  description: string;
  isMarked?: boolean;
  onDelete: VoidFunction;
  onEdit: VoidFunction;
  accessibilityLabel?: string;
};
