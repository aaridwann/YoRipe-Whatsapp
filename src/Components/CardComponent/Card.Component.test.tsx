import { render, fireEvent } from "@testing-library/react-native";

import CardComponent from "./Card.Component";
import { Props } from "./Card.Component.types";

jest.mock("@expo/vector-icons/Ionicons", () => "Ionicons");

beforeEach(() => jest.clearAllMocks());

describe("CardComponent", () => {
  const mockEdit = jest.fn();
  const mockDelete = jest.fn();
  const MockProps: Props = {
    title: "title",
    description: "description",
    onEdit: mockEdit,
    onDelete: mockDelete,
    accessibilityLabel: "notes",
  };

  describe("Snap test", () => {
    it("Should render with correct value", () => {
      const wrapper = render(<CardComponent {...MockProps} />);

      expect(wrapper).toMatchSnapshot();
    });
  });

  it("Should correct value according a props", () => {
    const { getByText, getAllByLabelText } = render(
      <CardComponent {...MockProps} />
    );
    const editButton = getAllByLabelText(
      `${MockProps.accessibilityLabel}_Edit_Button`
    );
    const deleteButton = getAllByLabelText(
      `${MockProps.accessibilityLabel}_Delete_Button`
    );

    expect(getByText(MockProps.title)).toBeTruthy();
    expect(getByText(MockProps.description)).toBeTruthy();
    expect(editButton).toBeTruthy();
    expect(deleteButton).toBeTruthy();
  });

  it("Should call Edit button while edit button is pressed", () => {
    const { getAllByLabelText } = render(<CardComponent {...MockProps} />);

    const editButton = getAllByLabelText(
      `${MockProps.accessibilityLabel}_Edit_Button`
    );

    fireEvent.press(editButton[0]);

    expect(MockProps.onDelete).not.toHaveBeenCalled();
    expect(MockProps.onEdit).toHaveBeenCalled();
  });

  it("Should call Edit button while edit button is pressed", () => {
    const { getAllByLabelText } = render(<CardComponent {...MockProps} />);

    const deleteButton = getAllByLabelText(
      `${MockProps.accessibilityLabel}_Delete_Button`
    );

    fireEvent.press(deleteButton[0]);

    expect(MockProps.onDelete).toHaveBeenCalled();
    expect(MockProps.onEdit).not.toHaveBeenCalled();
  });

  it("Should render Marked icon while data is Marked", () => {
    const newProps = {
      ...MockProps,
      isMarked: true,
    };
    const { getByLabelText } = render(<CardComponent {...newProps} />);

    const markedIcon = getByLabelText(
      `${MockProps.accessibilityLabel}_Marked_Icon`
    );

    expect(markedIcon).toBeTruthy();
  });
});
