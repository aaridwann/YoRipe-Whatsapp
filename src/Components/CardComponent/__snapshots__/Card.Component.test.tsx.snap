// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardComponent Snap test Should render with correct value 1`] = `
<View
  style={
    {
      "marginVertical": 10,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "white",
        "borderRadius": 8,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginVertical": 8,
        "minHeight": 130,
        "overflow": "hidden",
        "padding": 16,
        "width": 640,
      }
    }
  >
    <View
      style={
        {
          "height": "100%",
        }
      }
    >
      <Text
        accessibilityLabel="notes"
        style={
          {
            "fontWeight": "700",
          }
        }
      >
        title
      </Text>
      <Text
        style={
          {
            "marginTop": 10,
            "textAlign": "left",
          }
        }
      >
        description
      </Text>
    </View>
    <View
      style={
        {
          "flexDirection": "row",
        }
      }
    >
      <View
        accessibilityLabel="notes_Edit_Button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <Ionicons
          color="gray"
          name="pencil-outline"
          size={40}
          style={
            {
              "margin": 4,
              "marginTop": 20,
            }
          }
        />
      </View>
      <View
        accessibilityLabel="notes_Delete_Button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <Ionicons
          color="gray"
          name="trash-outline"
          size={40}
          style={
            {
              "margin": 4,
              "marginTop": 20,
            }
          }
        />
      </View>
    </View>
  </View>
</View>
`;
